src/avl.py

```diff
@@ -0,0 +1,468 @@
+import json
+import logging
+
+from hypothesis import given
+import hypothesis.strategies as st
+
+
+
+class AVLTree():
+    """The interface to an AVL tree.
+
+    The tree itself is actually represented by AVLTreeNode instances, but because of
+    needing to be able to change which node is the root, we need a container class.
+
+    Args:
+      value: an initial value to insert into the tree.
+    """
+
+    def __init__(self, value=None):
+        self._root = AVLTreeNode(self)
+        if value is not None:
+            self._root.insert(value)
+
+        self.log = []
+
+    def insert(self, value):
+        """Inserts the vallue into the tree.
+
+        Args:
+          value: the value to insert into the tree
+        """
+
+        self.root.insert(value)
+
+    def write_log(self, file_name="tree_log.json"):
+        """Writes a log of operation description and tree pairs in JSON for visualization."""
+
+        with open(file_name, "w") as f:
+            f.write(json.dumps(self.log))
+
+    def __str__(self):
+        return str(self.root)
+
+    @property
+    def root(self):
+        return self._root
+
+    @root.setter
+    def root(self, value):
+        self._root = value
+        self._root.parent = None
+
+    def to_dict(self):
+        """Returns a dictionary representation of this tree."""
+
+        return self._root.to_dict()
+
+    def __iter__(self):
+        return iter(self._root)
+
+
+class AVLTreeNode():
+    def __init__(self, container):
+        self.value = None
+        self._left = None
+        self._right = None
+        self.parent = None
+        self.container = container
+        self.left_height = self.right_height = 0
+
+    @property
+    def left(self):
+        return self._left
+
+    @property
+    def right(self):
+        return self._right
+
+    def __str__(self, depth=0):
+        def helper(child):
+            """Simplifies printing children."""
+
+            return "\n" + (child.__str__(depth + 1)
+                           if child is not None
+                           else "  " * (depth + 1) + "None")
+
+        return ("  " * depth + (
+            "AVLTreeNode("
+            "id = %s, "
+            "value = %s, "
+            "balance_factor = %s, "
+            "left_height = %s, "
+            "right_height = %s)"
+        ) % (id(self),
```

Comment:
consider using `format` for named interpolation: https://docs.python.org/2/library/string.html#format-examples